### PR TITLE
[23.1] make sure that TMP, TEMP, and TMPDIR are set

### PR DESCRIPTION
**Pull request by bernt-matthias**
seems needed for docker volume strings (an empty
variable would lead to `-v "::rw"` which apparently sometimes causes
`docker: invalid spec: ::rw: empty section between colons.`)

Not sure if this is the right place to add. In particular I was
wondering where environment variables from the destination
configuration are set.

I was not able to reproduce the problem locally. But at least
a test at my fork of the IUC repo has been successful: 
https://github.com/bernt-matthias/tools-iuc/pull/103

I guess we should test the whole IUC repo against this branch.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh [hunk 1]
```diff
@@ -39,6 +39,29 @@ GALAXY_LIB="$galaxy_lib"
...
+
+export TMP
+export TEMP
+export TMPDIR
+
+echo "_GALAXY_JOB_TMP_DIR $_GALAXY_JOB_TMP_DIR" >> outputs/tmp 
+echo "TMP $TMP" >> outputs/tmp
+echo "TEMP $TEMP" >> outputs/tmp 
+echo "TMPDIR $TMPDIR" >> outputs/tmp 
+
```
Comment (mvdbeek) [inline lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh]: what's that for ?

File: lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh [hunk 2]
```diff
@@ -39,6 +39,29 @@ GALAXY_LIB="$galaxy_lib"
...
+TMP="${TMP:-$TEMP}"
+TMPDIR="${TMPDIR:-$TEMP}"
+
+TMP="${TMP:-$TMPDIR}"
+TEMP="${TEMP:-$TMPDIR}"
+
+TMP="${TMP:-_GALAXY_JOB_TMP_DIR}"
+TEMP="${TEMP:-_GALAXY_JOB_TMP_DIR}"
+TMPDIR="${TMPDIR:-_GALAXY_JOB_TMP_DIR}"
```
Comment (mvdbeek) [inline lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh]:
```suggestion
TMP="${TMP:-$_GALAXY_JOB_TMP_DIR}"
TEMP="${TEMP:-$_GALAXY_JOB_TMP_DIR}"
TMPDIR="${TMPDIR:-$_GALAXY_JOB_TMP_DIR}"
```

File: lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh (hunk 1 above)
Comment (mvdbeek) [inline lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh]:
```suggestion
```
?

File: lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh [hunk 3]
```diff
@@ -39,6 +39,29 @@ GALAXY_LIB="$galaxy_lib"
...
+TMPDIR="${TMPDIR:-$_GALAXY_JOB_TMP_DIR}"
+
+export TMP
+export TEMP
+export TMPDIR
+
+echo "_GALAXY_JOB_TMP_DIR $_GALAXY_JOB_TMP_DIR" >> outputs/tmp 
+echo "TMP $TMP" >> outputs/tmp
+echo "TEMP $TEMP" >> outputs/tmp 
+echo "TMPDIR $TMPDIR" >> outputs/tmp
```
Comment (bernt-matthias) [inline lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh]:
```suggestion
```

File: lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh (hunk 1 above)
Comment (bernt-matthias) [reply lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh]: Debug code .. forgot to remove it :)

Comment (bernt-matthias) [reply lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh]: Not sure about the export?

Comment (mvdbeek) [reply lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh]: ~~Shouldn't be necessary, doesn't harm anything either~~ probably a good idea

**Review by mvdbeek — APPROVED** [review]
Awesome, thank you!

**Comment by mvdbeek** [comment]
> which apparently sometimes causes

it also started happening in our integration tests. I suppose the version of docker got updated on the worker images ?

**Comment by bgruening** [comment]
We see tests against 23.0 also failing. Do we need to get this into 23.0 or even earlier as well?

**Comment by bernt-matthias** [comment]
For CI we can just upgrade the used branch. Alternatively, we can backport, but how far?

**Comment by mvdbeek** [comment]
It is on 23.0 already, I included it in my already approved extra files dir PR ... but if a sneaky backport.

**Pull request merged**